File: fastLPR_py/tests/xl/test_xl_2d_kde.py
# ==============================================================================
# Cross-Language Test: XL-02 - 2D KDE vs MATLAB Reference
# ==============================================================================
# Test file: fastLPR_py/tests/xl/test_xl_kde_2d.py
# Created: 2025-01-07
#
# This test validates the Python cv_fastkde() implementation against MATLAB
# reference data for 2D kernel density estimation.
#
# Reference file: fastLPR/tests/refs/crosslang_e2e/ref_2d_kde.mat
# Contains:
#   - x: Input data (N x 2 matrix)
#   - hlist: Bandwidth candidates (k x 2 matrix)
#   - lcv.m: LCV scores for all bandwidths (k x 1 vector)
#   - h.selected: Selected bandwidth using 1-SE rule (1 x 2 vector)
#   - id1se: Index of selected bandwidth (1-based, MATLAB)
#   - elapsed: MATLAB execution time (scalar)
#
# Pass Criteria (per CLAUDE.md):
#   - BW MaxErr < 0.05 (bandwidth selection threshold for 2D)
#   - id1se matches (python 0-based == matlab id1se - 1)
#   - LCV RelErr < 0.01 (1% relative error)
#   - Speed ratio < 8x MATLAB time
# ==============================================================================

import time
from pathlib import Path

import numpy as np
import pytest
import scipy.io as sio

import fastlpr
from fastlpr import cv_fastkde

# Import tolerances from conftest
import sys
sys.path.insert(0, str(Path(__file__).parent.parent))
from conftest import TOL_CROSSLANG


def _get_project_root() -> Path:
    """Get project root (jss-code) from xl test directory."""
    return Path(__file__).parent.parent.parent.parent


def _get_ref_path() -> Path:
    """Get the path to MATLAB reference file for 2D KDE."""
    return _get_project_root() / "fastLPR" / "tests" / "refs" / "crosslang_e2e" / "ref_2d_kde.mat"


class TestXL02KDE2D:
    """Cross-language test: 2D KDE vs MATLAB reference."""

    @pytest.fixture
    def ref_data(self):
        """Load MATLAB reference data for 2D KDE."""
        ref_path = _get_ref_path()

        if not ref_path.exists():
            pytest.skip(f"MATLAB reference file not found: {ref_path}")

        ref = sio.loadmat(str(ref_path), squeeze_me=True, struct_as_record=False)
        return ref

    def test_2d_kde_matches_matlab(self, ref_data):
        """Test that 2D KDE matches MATLAB reference data."""
        ref = ref_data

        # Extract reference data
        x_ref = np.asarray(ref['x'])
        hlist_ref = np.asarray(ref['hlist'])

        # Ensure proper shapes
        if x_ref.ndim == 1:
            x_ref = x_ref.reshape(-1, 2)
        n = x_ref.shape[0]
        dx = x_ref.shape[1]

        if hlist_ref.ndim == 1:
            hlist_ref = hlist_ref.reshape(-1, dx)
        dh = hlist_ref.shape[0]

        # Extract MATLAB results
        lcv_m_ref = np.asarray(ref['lcv_m']).ravel()
        h_selected_ref = np.asarray(ref['h_selected']).ravel()
        id1se_ref = int(ref['id1se'])  # 1-based MATLAB index
        matlab_time = float(ref['elapsed'])

        # ====================================================================
        # Run Python implementation
        # ====================================================================
        start_time = time.perf_counter()
        kde = cv_fastkde(x_ref, h=hlist_ref, options={'verbose': False})
        py_time = time.perf_counter() - start_time

        # Extract Python results
        h_selected_py = np.asarray(kde.h).ravel()
        id1se_py = kde.lcv['id1se']  # 0-based Python index
        lcv_m_py = np.asarray(kde.lcv['lcv_m']).ravel()

        # ====================================================================
        # Compute error metrics
        # ====================================================================

        # Bandwidth selection error (max over dimensions)
        bw_maxerr = float(np.max(np.abs(h_selected_py - h_selected_ref)))

        # Index comparison (Python 0-based == MATLAB id1se - 1)
        id1se_match = (id1se_py == id1se_ref - 1)

        # LCV relative error: max(|py - ref| / |ref|)
        lcv_abs_err = np.abs(lcv_m_py - lcv_m_ref)
        lcv_rel_err = float(np.max(lcv_abs_err / np.abs(lcv_m_ref)))

        # Speed ratio
        speed_ratio = py_time / matlab_time

        # ====================================================================
        # Report results
        # ====================================================================
        print(f"\n=== XL-02: 2D KDE Cross-Language Test ===")
        print(f"Sample size (N): {n}")
        print(f"Dimensions: {dx}")
        print(f"Number of bandwidths: {dh}")
        print(f"MATLAB time: {matlab_time:.3f}s")
        print(f"Python time: {py_time:.3f}s")
        print(f"MATLAB selected h: [{h_selected_ref[0]:.6f}, {h_selected_ref[1]:.6f}] (id={id1se_ref})")
        print(f"Python selected h: [{h_selected_py[0]:.6f}, {h_selected_py[1]:.6f}] (id={id1se_py + 1})")
        print(f"\n--- Error Metrics ---")
        print(f"BW MaxErr: {bw_maxerr:.2e} (threshold: 0.05)")
        print(f"id1se match: {id1se_match} (py={id1se_py}, matlab-1={id1se_ref - 1})")
        print(f"LCV RelErr: {lcv_rel_err:.4%} (threshold: {TOL_CROSSLANG['lcv_relerr']:.0%})")
        print(f"Speed ratio: {speed_ratio:.2f}x (threshold: {TOL_CROSSLANG['speed_ratio']}x)")

        # ====================================================================
        # Assertions
        # ====================================================================

        # 1. Bandwidth abs error < 0.05 for 2D
        assert bw_maxerr < 0.05, (
            f"BW MaxErr ({bw_maxerr:.2e}) should be < 0.05"
        )

        # 2. id1se matches (python 0-based == matlab id1se - 1)
        assert id1se_match, (
            f"id1se mismatch: Python {id1se_py} != MATLAB {id1se_ref} - 1"
        )

        # 3. LCV rel err < TOL_CROSSLANG['lcv_relerr']
        assert lcv_rel_err < TOL_CROSSLANG['lcv_relerr'], (
            f"LCV RelErr ({lcv_rel_err:.4%}) should be < {TOL_CROSSLANG['lcv_relerr']:.0%}"
        )

        # 4. speed_ratio < TOL_CROSSLANG['speed_ratio']
        assert speed_ratio < TOL_CROSSLANG['speed_ratio'], (
            f"Speed ratio ({speed_ratio:.2f}x) should be < {TOL_CROSSLANG['speed_ratio']}x"
        )

        # ====================================================================
        # Summary
        # ====================================================================
        bw_pass = bw_maxerr < 0.05
        lcv_pass = lcv_rel_err < TOL_CROSSLANG['lcv_relerr']
        speed_pass = speed_ratio < TOL_CROSSLANG['speed_ratio']
        overall = bw_pass and id1se_match and lcv_pass and speed_pass

        print(f"\n--- Status ---")
        print(f"BW selection: {'PASS' if bw_pass else 'FAIL'}")
        print(f"id1se match: {'PASS' if id1se_match else 'FAIL'}")
        print(f"LCV RelErr: {'PASS' if lcv_pass else 'FAIL'}")
        print(f"Speed: {'PASS' if speed_pass else 'FAIL'}")
        print(f"Overall: {'PASS' if overall else 'FAIL'}")
        print("=" * 40)

    def test_2d_kde_produces_valid_density(self, ref_data):
        """Test that 2D KDE produces valid density estimates."""
        ref = ref_data

        x_ref = np.asarray(ref['x'])
        hlist_ref = np.asarray(ref['hlist'])

        if x_ref.ndim == 1:
            x_ref = x_ref.reshape(-1, 2)
        if hlist_ref.ndim == 1:
            hlist_ref = hlist_ref.reshape(-1, 2)

        kde = cv_fastkde(x_ref, h=hlist_ref, options={'verbose': False})

        # Density should be non-negative
        assert np.all(kde.fhat >= -1e-10), "All density values should be non-negative"

        # Density should integrate to approximately 1
        grid0 = kde.grid[0]
        grid1 = kde.grid[1]
        dx0 = grid0[1] - grid0[0]
        dx1 = grid1[1] - grid1[0]
        integral = float(np.sum(kde.fhat) * dx0 * dx1)

        assert abs(integral - 1.0) < 0.15, (
            f"Density integral ({integral:.4f}) should be close to 1"
        )

        # Selected bandwidth should be 2D
        assert len(kde.h) == 2, "Selected bandwidth should have 2 components"


if __name__ == "__main__":
    pytest.main([__file__, "-v", "-s"])

File: fastLPR_R/tests/testthat/test-xl-2d-kde.R
# Integration test for 2D KDE against MATLAB reference
# Task 4.2: Verify 2D KDE implementation matches MATLAB reference data
#
# 2D KDE performance (as of 2025-12-14):
# - Current ratio: ~2.88x MATLAB (EXCELLENT - target was < 3x!)
# - BW MaxErr: 0.00 (exact bandwidth match)
# - id1se: 17 (exact index match)
# - Status: PASS
#
# Note: This test monitors if Rcpp is being used, which is the key metric
# after optimization fixes. Even without Rcpp, pure R achieves < 3x ratio.
#
# Reference data structure (from ref_2d_kde.mat):
# - x: 2000 x 2 matrix (data points)
# - hlist: 20 x 2 matrix (bandwidth candidates)
# - h.selected: [0.6951928, 0.6951928]
# - id1se: 17
# - elapsed: 0.1249718 seconds (MATLAB time)
# - lcv.m: 20 values (LCV scores for each bandwidth)

context("Cross-validation: 2D KDE against MATLAB")

test_that("2D KDE matches MATLAB reference", {
  # Skip if MATLAB reference data not available (CRAN compatibility)
  skip_if_no_matlab_refs()

  # Load R.matlab for reading .mat files
  suppressWarnings(library(R.matlab))  # Suppress version mismatch warning

  # Find project root
  pkg_root <- getwd()
  while (!grepl("fastLPR_R$", basename(pkg_root)) && dirname(pkg_root) != pkg_root) {
    pkg_root <- dirname(pkg_root)
  }
  project_root <- dirname(pkg_root)

  # Reference file path
  ref_path <- file.path(project_root, "fastLPR", "tests", "refs", "crosslang_e2e", "ref_2d_kde.mat")

  # Skip if reference file not found
  skip_if_not(file.exists(ref_path),
              message = paste("Reference file not found:", ref_path))

  # Report Rcpp status (KEY METRIC for optimization)
  rcpp_status <- FALSE
  if (exists("rcpp_available") && is.function(rcpp_available)) {
    rcpp_status <- rcpp_available()
    if (rcpp_status && exists("get_rcpp_info") && is.function(get_rcpp_info)) {
      info <- tryCatch(get_rcpp_info(), error = function(e) NULL)
      if (!is.null(info)) {
        message(sprintf("Rcpp: ENABLED, %d threads, OpenMP: %s",
                        info$num_threads,
                        ifelse(info$num_threads > 1, "YES", "NO")))
      } else {
        message("Rcpp: ENABLED (info unavailable)")
      }
    } else if (!rcpp_status) {
      message("Rcpp: NOT available - 2D KDE using pure R fallback")
    }
  } else {
    message("Rcpp: Status check function not available")
  }

  # Load MATLAB reference data
  mat <- readMat(ref_path)

  # Extract inputs
  x <- as.matrix(mat$x)
  hlist <- as.matrix(mat$hlist)
  n <- nrow(x)
  dx <- ncol(x)

  # Verify dimensions
  expect_equal(dx, 2, label = "Expected 2D data")
  message(sprintf("Test data: N=%d, dx=%d, hlist size=%d x %d",
                  n, dx, nrow(hlist), ncol(hlist)))

  # Get MATLAB timing
  matlab_time <- if (!is.null(mat$elapsed)) as.numeric(mat$elapsed) else NA
  message(sprintf("MATLAB reference time: %.3fs", matlab_time))

  # Configure options (matching MATLAB defaults)
  opt <- list(
    order = 0,       # KDE
    calc_dof = FALSE
    # Note: flag_power2 defaults to TRUE for 2D (unlike 3D where it's FALSE)
  )

  # Time the R execution
  r_time <- system.time({
    result <- cv_fastkde(x, hlist, opt)
  })[["elapsed"]]

  # Calculate speed ratio
  ratio <- if (!is.na(matlab_time) && matlab_time > 0) r_time / matlab_time else NA

  # Report timing
  message(sprintf("2D KDE: R=%.3fs, MATLAB=%.3fs, Ratio=%.2fx",
                  r_time,
                  ifelse(is.na(matlab_time), 0, matlab_time),
                  ifelse(is.na(ratio), 0, ratio)))

  # Compare bandwidth selection (REQ-5: All metrics must pass)
  m_bw <- as.vector(mat$h.selected)
  r_bw <- as.vector(result$h)

  bw_maxerr <- max(abs(r_bw - m_bw))
  message(sprintf("Bandwidth: R=[%.4f,%.4f], MATLAB=[%.4f,%.4f]",
                  r_bw[1], r_bw[2], m_bw[1], m_bw[2]))
  message(sprintf("BW MaxErr: %.2e", bw_maxerr))

  # Bandwidth selection MUST match
  # Threshold: 0.05 for 2D
  expect_lt(bw_maxerr, 0.05,
            label = sprintf("2D KDE BW MaxErr (%.4f)", bw_maxerr))

  # Compare selected bandwidth indices
  if (!is.null(mat$id1se) && !is.null(result$lcv$id1se)) {
    m_id1se <- as.integer(mat$id1se)
    r_id1se <- result$lcv$id1se
    message(sprintf("id1se: R=%d, MATLAB=%d, Match=%s",
                    r_id1se, m_id1se, ifelse(r_id1se == m_id1se, "YES", "NO")))

    # Index should match exactly
    expect_equal(r_id1se, m_id1se,
                 label = sprintf("id1se: R=%d, MATLAB=%d", r_id1se, m_id1se))
  }

  # Compare LCV values using RELATIVE error per CLAUDE.md 8.4
  # For KDE (order=0): LCV RelErr < 0.01 (1%)
  # Note: Absolute errors can be large (e.g., 0.184) when LCV values are large (e.g., -6228)
  # but relative error is what matters for log-likelihood comparisons
  if (!is.null(mat$lcv.m) && !is.null(result$lcv$lcv_m)) {
    m_lcv <- as.vector(mat$lcv.m)
    r_lcv <- result$lcv$lcv_m

    # Ensure same length
    min_len <- min(length(r_lcv), length(m_lcv))
    lcv_abs_err <- abs(r_lcv[1:min_len] - m_lcv[1:min_len])
    lcv_maxerr <- max(lcv_abs_err)

    # Compute RELATIVE error: max(|r - m| / |m|)
    lcv_rel_err <- max(lcv_abs_err / abs(m_lcv[1:min_len]))

    message(sprintf("LCV MaxErr: %.2e (abs), RelErr: %.4f%% (threshold: < 1%%)",
                    lcv_maxerr, lcv_rel_err * 100))

    # Per CLAUDE.md 8.4: KDE uses LCV RelErr < 0.01 (1%) threshold
    expect_lt(lcv_rel_err, 0.01,
              label = sprintf("LCV RelErr (%.4f%%)", lcv_rel_err * 100))
  }

  # Speed ratio check (target < 8x for JSS, ideally < 3x after optimization)
  if (!is.na(ratio)) {
    # Report performance category with Rcpp status
    if (ratio < 1.0) {
      perf_status <- "FASTER than MATLAB"
    } else if (ratio < 3.0) {
      perf_status <- "EXCELLENT (< 3x - target achieved)"
    } else if (ratio < 5.0) {
      perf_status <- "GOOD (< 5x)"
    } else if (ratio < 8.0) {
      perf_status <- "ACCEPTABLE (< 8x JSS threshold)"
    } else if (ratio < 15.0) {
      perf_status <- "NEEDS OPTIMIZATION (8-15x)"
    } else {
      perf_status <- "CRITICAL (> 15x)"
    }

    message(sprintf("Performance: %s", perf_status))

    # For JSS publication, ratio must be < 8x
    expect_lt(ratio, 8,
              label = sprintf("Speed ratio (%.2fx)", ratio))

    # Additional warning if not meeting target
    if (ratio >= 3.0) {
      message(sprintf("NOTE: Ratio %.2fx > 3x target. Rcpp: %s",
                      ratio, ifelse(rcpp_status, "enabled", "disabled")))
    }
  }

  # Verify output structure
  expect_true(is.list(result))
  expect_true(!is.null(result$fhat))
  expect_true(!is.null(result$h))
  expect_true(!is.null(result$xlist))
  expect_equal(length(result$h), 2)
  expect_equal(length(result$xlist), 2)

  # Check density is non-negative
  expect_true(all(result$fhat >= 0), label = "Density non-negative")
})

# ARCHIVED: 2026-01-09 - "2D KDE basic functionality works with test data" (moved to dev/archive/tests-archive-20260109/r/xl/archived_test-xl-2d-kde.R)


test_that("2D KDE with mixture distribution selects appropriate bandwidth", {
  # Skip if functions not loaded
  skip_if_not(exists("cv_fastkde") && is.function(cv_fastkde),
              message = "cv_fastkde not available")
  skip_if_not(exists("get_hlist") && is.function(get_hlist),
              message = "get_hlist not available")

  # Test bandwidth selection on bimodal 2D data
  set.seed(42)
  n1 <- 100
  n2 <- 100

  # Two well-separated clusters
  x1 <- cbind(rnorm(n1, -2, 0.5), rnorm(n1, -2, 0.5))
  x2 <- cbind(rnorm(n2, 2, 0.5), rnorm(n2, 2, 0.5))
  x <- rbind(x1, x2)

  hlist <- get_hlist(c(7, 7), list(c(0.1, 1.5), c(0.1, 1.5)))

  kde <- cv_fastkde(x, hlist, list(verbose = FALSE, N = c(50, 50)))

  # Bandwidth should be selected
  expect_true(all(kde$h >= 0.1))
  expect_true(all(kde$h <= 1.5))

  # Should have LCV results
  expect_true(!is.null(kde$lcv))
  expect_true(!is.null(kde$lcv$lcv_m))
})

# ARCHIVED: 2026-01-09 - "2D KDE reproducibility check" (moved to dev/archive/tests-archive-20260109/r/xl/archived_test-xl-2d-kde.R)


# ARCHIVED: 2026-01-09 - "2D KDE Rcpp acceleration detection" (moved to dev/archive/tests-archive-20260109/r/xl/archived_test-xl-2d-kde.R)


# ARCHIVED: 2026-01-09 - "2D KDE density integrates to approximately 1" (moved to dev/archive/tests-archive-20260109/r/xl/archived_test-xl-2d-kde.R)


# ARCHIVED: 2026-01-09
# Source: fastLPR_R/tests/testthat/test-xl-2d-kde.R
# Archive: dev/archive/tests-archive-20260109/r/xl/archived_test-xl-2d-kde.R
# Archived tests:
# - "2D KDE basic functionality works with test data"
# - "2D KDE reproducibility check"
# - "2D KDE Rcpp acceleration detection"
# - "2D KDE density integrates to approximately 1"
